ships/0025-strategy-parameters-enhancements.md

```diff
@@ -0,0 +1,281 @@
+- Allow build users to define values for a string parameter, or an element of an array parameter, using a ConfigMap key reference.
+- Allow build users to define a prefix for the Secret value.
+- Allow build users to define a prefix for a ConfigMap value.
+
+### Non-Goals
+
+## Proposal
+
+### API Changes
+
+We will extend the Build Strategy Parameter type to support the following:
+
+```yaml
+spec:
+  parameters:
+    - name: build-args
+      description: The values for the ARGs in the Dockerfile.
+      type: array
+      defaults: []
+```
+
+The `type` property is new and allows either `string` or `array` with `string` being the default.
+
+The `defaults` parameter is new and is a string array. No possibility here to reference Secret or ConfigMap values.
+
+The Build and BuildRun parameterValue will be extended to support the following:
+
+```yaml
+spec:
+  paramValues:
+    - name: single-secret-value
+      secretValue:
+        name: secret-name
+        key: secret-key
+    - name: single-configmap-value
+      configMapValue:
+        name: configmap-name
+        key: configmap-key
+    - name: simple-list
+      values:
+        - value: first-item
+        - value: second-item
+    - name: build-args
+      values:
+        - value: KEY1=myarg1
+        - secretValue:
+            name: build-arg-secret
+            key: value
+            format: "KEY2=${SECRET_VALUE}" 
+        - configMapValue:
+            name: build-configuration
+            key: base-image-digest
+            format: "BASE_IMAGE_DIGEST=${CONFIGMAP_VALUE}"
+```
+
+The `secretValue` property is new and can be used instead of value to reference a key in a Secret. Optionally, the `format` property (shown above at the end inside an array) can be used to add text around the Secret value. The scenario is clear there: the value for the build arg comes from a Secret, but the key of the build arg also needs to go there.
+
+The `configMapValue` property is also new and works in the same way as the `secretValue`. A `format` is also supported here.
+
+The `values` property is new and allows to specify multiple values. Each value can either be a simple string `value`, or a `secretValue`, or a `configMapValue`.
+
+Validation will be added ...
+
+- to the Build and BuildRun to verify that the value for an array parameter is not a single value, and vice versa.
+- to the Build and BuildRun to verify the existence of the Secret and the referenced key.
+- to the Build and BuildRun to verify the existence of the ConfigMap and the referenced key.
+
+### Backend
+
+We will continue to use Tekton results. This is straight-forward for the array support.
+
+For Secrets, we will do the following:
+
+- For each `secretKeyValue`, we will add an environment variable to those steps that reference the param.
+  - The environment variable name will be `SHP_SECRET_PARAM_{randomSuffix}`
+  - The env variable on the step will have a secretKeyRef.
```

Comment:
Assume you mean https://github.com/shipwright-io/build/blob/ac139dba9437bca165137856e18a593417945630/vendor/k8s.io/api/core/v1/types.go#L1946 

providing a reference link to avoid confusion, and not require the reader to remember all the type structure of k8s env  vars (which is not trivial IMO) would be nice

Reply:
Good idea, adding links.